### PR TITLE
Add test-sos as a submodule

### PR DESCRIPTION
**Pull request by kholland-intel**

### CODE REVIEW

File: autogen.sh
Comment (davidozog) [inline autogen.sh:5]: The warning output doesn't look right to me on my system.  It's ok to have a few instances of `echo` to force the newlines.

Also, let's just `exit 1` if the submodule is not there because the whole build will fail in that case as things are now.

Comment (kholland-intel) [reply autogen.sh:5]: If we're going to be exiting should we also upgrade the "Warning" to an "Error"?

Comment (davidozog) [reply autogen.sh:5]: Yes please, thanks.

**Comment by davidozog** [comment]
Let's add `--recursive-submodule` to the example build scripts.

File: .github/workflows/ci.yml
Comment (davidozog) [inline .github/workflows/ci.yml:237]:
```suggestion
          # RPM build does not handle the symbolic link, so manually copy tests-sos/test:
          rm test
```
A comment here is probably helpful to guide whoever may need to generate an RPM in the future

**Review by davidozog — DISMISSED** [review]
Looks good!  Suggest adding a comment in the RPM build in CI.

File: .gitmodules
Comment (philipmarshall21) [inline .gitmodules:4]: Should this point to main or a tagged commit? Should be fine, so long as we are ok with any changes made to main being pulled by users. Is there a possibility while working on changes for OpenSHMEM 1.x.y we would add tests to the main branch on tests-sos that we wouldn't necessarily want users to pull when building the test-sos repo with an older OpenSHMEM version. Please let me know if I didn't explain that well 😅

Comment (davidozog) [reply .gitmodules:4]: Great question.  I prefer `main` align with the `main` of SOS, because the addition of a new test (for example) should continue to be tested throughout development cycle towards the next release.  However, when we do release we will likely want to point to a tag or branch here, so this should be added to our release checklist.

I think we should document how to handle things like this in the SOS wiki before merging this PR.  As another example, developers need to be aware that this change will likely cause some pull requests to have 2 parts: one for SOS and one for tests-sos.  This could be trickier than it sounds, because the SOS part will need to test against the tests-sos part, likely before the tests-sos part is merged...

File: .github/workflows/ci.yml
Comment (wrrobin) [inline .github/workflows/ci.yml:672]: Why `v4` not `v2`?

File: autogen.sh
Comment (wrrobin) [inline autogen.sh:10]: A naive question.. is there a way to auto download it?

**Comment by wrrobin** [comment]
I did not try it out, but do we have to use `recurse-submodule` now every time? Do we need any update on our README or wiki pages?

File: .github/workflows/ci.yml
Comment (davidozog) [reply .github/workflows/ci.yml:672]: We tentatively decided to update because it seems there have been bug fixes in the checkout actions since `v2` that could affect submodules:

https://github.com/actions/checkout/blob/main/CHANGELOG.md

But I'm not sure we really _need_ to update... but it's working so I don't see the harm.

Does that answer the question, or do you have a concern about updating @wrrobin?

File: autogen.sh
Comment (davidozog) [reply autogen.sh:10]: We could definitely auto download it here in `autogen.sh` but based on a few other projects (e.g. ompi, mpich, oshmpi, raspberry pi, etc) it seems this is not standard practice.  Possibly because users (developers, primarily) may not always want to `update --init` submodules when running `autogen.sh`.  On the other hand, if no files are there, maybe we should do it automatically, but seems like it's probably best to let a human figure it out.  Thoughts?

**Comment by davidozog** [comment]
> I did not try it out, but do we have to use `recurse-submodule` now every time? Do we need any update on our README or wiki pages?

Yes, if you want to build SOS successfully you will now need to call `git clone --recurse-submodules <SOS>` to clone _both_ SOS and the `tests-sos` submodule.  You could alternatively clone normally (i.e. without `--recurse-submodules`) then run `git submodule update --init` from inside the repo before running `autogen.sh`.

I think we should also consider a feature where SOS can be built _without_ having to clone and build the tests, but that is not supported in this PR.  Do you think we should do that before merging this change?

I absolutely think we should document everything before merging.  Also please note my comment above about testing any PR that changes both SOS and the tests-sos submodule...  This procedure could be confusing without documentation.

File: autogen.sh
Comment (wrrobin) [reply autogen.sh:10]: I see the reason.. a user may not want to auto download in case there is a local version there already downloaded and possibly updated with newer changes. But, as you also mentioned, for the latter case (no files are there), we can do it auto, right? Is there any drawbacks or cases where it may not help?

**Comment by wrrobin** [comment]
@davidozog - I was trying this branch out and the submodule download always getting stuck for me. Not sure if the issue is related with git config. Did you see this?
```
$> git clone -b "pr/test-submodule" --recurse-submodules https://github.com/kholland-intel/SOS.git sos-submodule
Cloning into 'sos-submodule'...
remote: Enumerating objects: 17958, done.
remote: Counting objects: 100% (1362/1362), done.
remote: Compressing objects: 100% (499/499), done.
remote: Total 17958 (delta 930), reused 1162 (delta 825), pack-reused 16596
Receiving objects: 100% (17958/17958), 5.33 MiB | 18.31 MiB/s, done.
Resolving deltas: 100% (13463/13463), done.
Updating files: 100% (1806/1806), done.
Submodule 'modules/tests-sos' (git@github.com:openshmem-org/tests-sos.git) registered for path 'modules/tests-sos'
Cloning into '/mnt/scratch/rahmanmd/sos-submodule/modules/tests-sos'...
```

**Comment by davidozog** [comment]
> @davidozog - I was trying this branch out and the submodule download always getting stuck for me. Not sure if the issue is related with git config. Did you see this?
> 
> ```
> $> git clone -b "pr/test-submodule" --recurse-submodules https://github.com/kholland-intel/SOS.git sos-submodule
> Cloning into 'sos-submodule'...
> remote: Enumerating objects: 17958, done.
> remote: Counting objects: 100% (1362/1362), done.
> remote: Compressing objects: 100% (499/499), done.
> remote: Total 17958 (delta 930), reused 1162 (delta 825), pack-reused 16596
> Receiving objects: 100% (17958/17958), 5.33 MiB | 18.31 MiB/s, done.
> Resolving deltas: 100% (13463/13463), done.
> Updating files: 100% (1806/1806), done.
> Submodule 'modules/tests-sos' (git@github.com:openshmem-org/tests-sos.git) registered for path 'modules/tests-sos'
> Cloning into '/mnt/scratch/rahmanmd/sos-submodule/modules/tests-sos'...
> ```
> 
> Quick google search led to https://stackoverflow.com/questions/40841882/automatically-access-git-submodules-via-ssh-or-https/44630028#44630028 which suggests to use relative path for submodules. Not sure if that is the fix.

Hi @wrrobin - this command worked on my end...  I wonder if you have github SSH authentication setup appropriately.  Are you able to do a test like this?
```
dave@wsl in ~ $ ssh -T git@github.com
Enter passphrase for key '/home/dave/.ssh/github':
Hi davidozog! You've successfully authenticated, but GitHub does not provide shell access.
```

Another option might be to change `.gitmodules` to track the https URL instead of the SSH... In fact, that seems much more common on other projects (e.g., OSHMPI, mpich, UCX, raspberry-pi, etc).

I usually see relative paths for projects hosted under the same organization, but there is an exception with PRRTE so I think it should work (across `Sandia-OpenSHMEM` and `openshmem-org`).  I think I prefer the relative path (if it works) because then the user chooses how to authenticate.  I will try this unless you prefer https or ssh?

**Comment by wrrobin** [comment]
The change worked fine. 

Do you need to have any changes for the `make` to work?

**Comment by davidozog** [comment]
> The change worked fine.

Great!  Yes, for me too.

> Do you need to have any changes for the `make` to work?

I don't think any changes are needed assuming the tests-sos submodule is initialized.  Running `make check` after cloning with `--recurse-submodules` works fine on my end.  A next step is to enable building SOS without having to clone the tests-sos submodule, but I think that's better handled on a separate PR.

So this is ready for review.  I assigned the PR to myself offload it from @kholland-intel, and will wait for approvals from you and @philipmarshall21.

**Comment by davidozog** [comment]
@wrrobin @philipmarshall21 - It seems like we're spending too much time waiting on the UCX with pmi-mpi test row.  I propose we remove it or swap it out with something else until it's resolved.

Is the UCX with pmi-mpi test case important to anybody listening?  @jdinan perhaps?  It's not needed on our end...

**Comment by wrrobin** [comment]
@davidozog - The changes resolved the issue of `make check`.

**Comment by davidozog** [comment]
> Tested with different build directories. Works perfectly.

This reminds me we need CI to cover the case of users building in the SOS root directory - added issue #1111 so we can merge this asap since it's holding up other PRs.

**Review by philipmarshall21 — APPROVED** [review]

**Pull request merged**
